### ISSUE TITLE
OnlyOffice can write changes through an "upload" public link share

### ISSUE THREAD

**Issue by kulmann** [owncloud/ocis#8609]
## Describe the bug
OnlyOffice can write documents from a public link share with "Can upload" role.

## Steps to reproduce
1. As "admin", create `test.docx` via OnlyOffice
2. As "admin", create a public link to this file with `Can upload` role (which states permissions `view, download and upload`)
3. As anonymous, receive and follow the public link
4. As anonymous, open `test.docx` via OnlyOffice
5. As anonymous, write some content into the file and leave the document again
6. As "admin", see that the changes have been written to disk (e.g. by downloading the file after the lock has been released)

## Expected behavior
"anonymous" should have a read only view or at least an error message when trying to save.

## Actual behavior
"anonymous" can edit and save the document.

**Comment by kulmann** [comment]
confirmed in both `v5.0.0-rc.5` and `v4.0.6`

**Comment by 2403905** [comment]
Step 2. Is `admin` create a public link to this file with the `Can upload ` role using API or UI ? 
The `Can upload ` role has to be used for folders, bu for files.

**Comment by phil-davis** [comment]
To clarify, the "upload" permission should allow "anonymous" to create new files in the folder, but not overwrite (=edit) existing files.

After "anonymous" has uploaded a file, they should not be able to change what they uploaded (can't edit it or delete it), but should be able to list the files and download files.

Correct?

**Comment by individual-it** [comment]
What API is OnlyOffice using for that? I'm surprised the API tests did not catch it

**Comment by micbar** [comment]
Wopi

**Comment by kulmann** [comment]
> To clarify, the "upload" permission should allow "anonymous" to create new files in the folder, but not overwrite (=edit) existing files.
> 
> After "anonymous" has uploaded a file, they should not be able to change what they uploaded (can't edit it or delete it), but should be able to list the files and download files.
> 
> Correct?

Correct, thank you!

**Comment by ScharfViktor** [comment]
We have e2e coverage for an app provider. It is not very large, because we wanted to implement only primitive tests as a first step (create and open a `docx` or `odt` document).
Maybe we should increase the coverage with e2e?

**Comment by individual-it** [comment]
Yes I think we should. I believe the decision to make only small coverage was because the security related code path should be the same as with webDAV, but that does not look like its true

**Comment by 2403905** [comment]
@kulmann I found the place in a code when we give the write permissions to OnlyOffice but it is not a root issue. Because the permissions statements (view, download and upload) do not contradict that.
If a user can upload then the user can change the file, right?

**Comment by 2403905** [comment]
@micbar @kulmann  Is the `upload` permission redundant in this case?

**Comment by kulmann** [comment]
> If a user can upload then the user can change the file, right?

No... @phil-davis interpretation is what we want: The user can upload files, but not overwrite existing files.

**Comment by 2403905** [comment]
We should clarify the public permission definitions.
FYI @tbsbdr @micbar

**Comment by individual-it** [comment]
How does the WebDAV API currently react in that case?

**Comment by tbsbdr** [comment]
Currently I see 3 options

1. fix the bug
2. Add "edit" to the role description
3. remove the whole role "Uploader"

I'd favour the 3rd option for now, as Uploader and Editor would then be almost the same. what do you think @kulmann @micbar ?

**Comment by kulmann** [comment]
Option 3 would be fine by me. UX-wise the differentiation is hard to understand anyway. So might actually be an improvement...

**Comment by micbar** [comment]
yes. Option 3 is the best.

**Comment by tbsbdr** [comment]
ok, so lets remove this role from the link 😬

**Comment by 2403905** [comment]
Is the "can upload"  should be changed to "can edit" or "can view"?

**Comment by micbar** [comment]
Yes. We need a better description which matches our behavior

**Comment by micbar** [comment]
@tbsbdr @kulmann 
Let us fix the description.

**Comment by tbsbdr** [comment]
@tbsbdr propose a text

**Comment by tbsbdr** [comment]
🙈 adding "edit" to the description won't work, as text or md files can not be edited.
@micbar @kulmann I currently see 2 options and would choose the one with least effort. do you agree?

1. fix the bug
2. ~Add "edit" to the role description~
3. remove the whole role "Uploader"

**Comment by tbsbdr** [comment]
I could Not Edit txt, can you?

**Comment by micbar** [comment]
you can upload a new txt file version.

**Comment by micbar** [comment]
The main problem here is the "edit" vs. "upload" semantics. From the ocis backend perspective, both are uploads and require the upload permission. This is different for directories. They are controlled by the create-directories permission.

**Comment by phil-davis** [comment]
Note: "edit-save" and "upload-overwrite" really mean the same thing. There is no such thing as "editing" a file that is sitting on a server. Actually, there is also no such thing as "editing" a file on the local filesystem on my laptop. "editing" is a thing that is done in-memory with a user-interface (graphical (LibreOffice...) or textual (VI, nano, a line-based command-line editor...)

When the person doing the "editing" asks to "save", then the editor software does an "upload-overwrite" to the file, wherever the persistent storage is. It might rewrite the whole file, it might append to the end (if the change just adds to the file), it might overwrite just some bytes of the file (if it implements some optimized way of saving just the bits that changed).

**Comment by JustKiddingCode** [comment]
From a linux file system perspective, which is similar to webdav:
"upload" rights should mean  "write permissions to directory, but not the files".
Uploading a file should check if the file exists and the user has write permissions _to that file_.

```bash
mkdir test
echo "bar" > test/foo # works, because of write permissions to test
chmod -w > test/foo
echo "foo" > test/foo # fails, permission denied
```

**Comment by tbsbdr** [comment]
@tbsbdr adjust wording for the role

**Comment by tbsbdr** [comment]
## Suggestion for the rewording of the roles description.

- Assuming that "Only for invited ppl" will be removed
- Adds `edit` to `Can upload`
- Adds `delete` to `Can edit`
- I'd prefer to be less verbose in the role description (option 2)
- Note: In general I'd prefer to **remove the Uploader role at all**, as it adds little value for the user, but if I understand @micbar correctly this would be expensive due to a migration necessity.

@kulmann is option 2 - less verbose, fine for you?

| 1) Verbose | 2) Less verbose |
|----|----|
| ![Image](https://github.com/user-attachments/assets/d839f86e-d3d7-43cd-ae68-62e295726038) | ![Image](https://github.com/user-attachments/assets/e666836a-aeb9-48ba-9818-f6200d353aa3) |


## General overview


|                  | Needs authentication | delete | upload/edit | view | download | upload (existing content not   revealed) |
|:----------------:|:--------------------:|:------:|:-----------:|:----:|:--------:|:----------------------------------------:|
|  Invited People  |           ✅          |   (✅)  |     (✅)     |  (✅) |    (✅)   |                     ❌                    |
|     Can edit     |           ❌          |    ✅   |      ✅      |   ✅  |     ✅    |                     ❌                    |
|    Can upload    |           ❌          |    ❌   |      ✅      |   ✅  |     ✅    |                     ❌                    |
|     Can view     |           ❌          |    ❌   |      ❌      |   ✅  |     ✅    |                     ❌                    |
| Secret File Drop |           ❌          |    ❌   |      ❌      |   ❌  |     ❌    |                     ✅                    |

**Comment by tbsbdr** [comment]
@kulmann whats your opinion - go or no? 👆

**Comment by kulmann** [comment]
We had that in the past and changed it to being more verbose because people had difficulties understanding the implications. Now that the whole context is more verbose we may as well go back to less verbose within the dropdown. Fine by me.

**Comment by rhafer** [comment]
@kulmann So I guess this would be a web ticket now? On the server side we don't have any descriptive texts for the public link types AFAICS.
